02-ADTs/homework-02/hw-02.hs

```diff
@@ -0,0 +1,198 @@
+failed disk, and a bored yet incompetent programmer, the log messages are horribly out of order. Until we do some organizing, there
+will be no way to make sense of what went wrong! We’ve designed a
+data structure that should help—a binary search tree of LogMessages:
+data MessageTree = Leaf
+| Node MessageTree LogMessage MessageTree
+Note that MessageTree is a recursive data type: the Node constructor itself takes two children as arguments, representing the left and
+right subtrees, as well as a LogMessage. Here, Leaf represents the
+empty tree.
+A MessageTree should be sorted by timestamp: that is, the timestamp of a LogMessage in any Node should be greater than all timestamps of any LogMessage in the left subtree, and less than all timestamps of any LogMessage in the right child.
+cis 194: homework 2 4
+Unknown messages should not be stored in a MessageTree since
+they lack a timestamp.
+-}
+
+{-
+Define a function
+insert :: LogMessage -> MessageTree -> MessageTree
+which inserts a new LogMessage into an existing MessageTree, producing a new MessageTree. insert may assume that it is given a
+sorted MessageTree, and must produce a new sorted MessageTree
+containing the new LogMessage in addition to the contents of the
+original MessageTree.
+However, note that if insert is given a LogMessage which is
+Unknown, it should return the MessageTree unchanged.
+-}
+--Exercise 2
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _)              tree                        = tree
+insert _                        tree@(Node _ (Unknown _) _) = tree
+insert msg                      Leaf                        = Node Leaf msg Leaf
+insert lm1@(LogMessage _ xts _) (Node left lm2@(LogMessage _ yts _) right)
+  | xts < yts = Node (insert lm1 left) lm2 right
+  | otherwise = Node left lm2 (insert lm1 right)
+
+
+{-
+Once we can insert a single LogMessage into a MessageTree,
+we can build a complete MessageTree from a list of messages. Specifically, define a function
+build :: [LogMessage] -> MessageTree
+which builds up a MessageTree containing the messages in the list,
+by successively inserting the messages into a MessageTree (beginning
+with a Leaf).
+-}
+--Exercise 3
+build :: [LogMessage] -> MessageTree
+build = foldr insert Leaf
+
+
+{-
+Finally, define the function
+inOrder :: MessageTree -> [LogMessage]
+which takes a sorted MessageTree and produces a list of all the
+LogMessages it contains, sorted by timestamp from smallest to biggest.
+(This is known as an in-order traversal of the MessageTree.)
+With these functions, we can now remove Unknown messages and
+sort the well-formed messages using an expression such as:
+inOrder (build tree)
+[Note: there are much better ways to sort a list; this is just an exercise to get you working with recursive data structures!]
+-}
+--Exercise 4
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf                  = []
+inOrder (Node left msg right) = inOrder left ++ [msg] ++ inOrder right
```

Comment:
This solution is correct, but inefficient because of all the list concatenation operations that have to be performed. Can you think of a more efficient solution that uses only the `:` operator to construct the list? 

Hint: You might need a helper function that takes an additional argument.

Reply:
@peironggg, this is correct! The solution I had in mind was something like this: 

```haskell
inOrder' :: MessageTree -> [LogMessage]
inOrder' = traverse []
  where
      traverse :: [LogMessage] -> MessageTree -> [LogMessage]
      traverse ms Leaf = ms
      traverse ms (Node left msg right) = traverse (msg : traverse ms right) left

```

If you inline the lambda that you pass to the `func` parameter in your solution, then you end up with my solution above:

```haskell
... 
foldTree initial Leaf = initial
foldTree initial (Node left msg right) = foldTree (msg : (foldTree initial right)) left
```

You have thus gone one step further and abstracted out the fold pattern (https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Foldable.html#t:Foldable), so good job!

Reply:
2 last comments on style:
- Expressions defined in the `where` clause should be placed on the next line, with a 2 space indentation from the start of the `where` keyword.
- Add the type annotations for the expressions in the where clause as well. 

See the following example: 

```haskell
foo :: a -> b -> Int
foo a b = getInt 
  where
    getInt :: Int
    getInt = 3
```

Reply:
@oishikg Awesome, thanks! 
Regarding the `foo` example provided above, is it a 2 space indentation to get to the `where` (given above) or a 4 space indentation as shown below?

```haskell
foo :: a -> b -> Int
foo a b = getInt 
    where
      getInt :: Int
      getInt = 3
```

Reply:
The `where` keyword (as an exception) takes a 2 space indentation. 

Here's the style guide we (more or less) try to stick to: 
[kowainik style guide](https://kowainik.github.io/posts/2019-02-06-style-guide)

Reply:
@oishikg Thankss